src/execution_plan/execution_plan_reduce_to_apply.c

```diff
@@ -0,0 +1,164 @@
+	ExecutionPlan_BindPlanToOps(branch_match_root, plan);
+	OpBase *argument = ExecutionPlan_LocateFirstOp(branch_match_root, OPType_ARGUMENT);
+	if(argument) {
+		if(argument->modifies) {
+			/* Don't lose information when optimizing this branch. The optimizations are lookgin for
+			 * specific operation types and if the result of the operations is achievable by one of its
+			 * decedents, this op is redundant. The argument op "modifies" the execution plan by
+			 * showing the bounded arguments. Given that, once optimized the label scan op is
+			 * so called redundant and will be removed, since its result is achievable by the argument op.
+			 * But this is wrong since the argument op is not modifing anything.
+			 * The next sequence should avoid the removal of the label scan op, and yet optimize the rest
+			 * of the branch. */
+			if(argument->parent->type == OPType_NODE_BY_LABEL_SCAN) {
+				OpBase *parent = argument->parent;
+				// Temporary remove label scan op. Now argument op is the top op.
+				ExecutionPlan_RemoveOp(plan, parent);
+				// Optimize the rest of the plan. Unfortunately, this currently needs to be done in place.
+				optimizePlan(match_branch_plan);
+				// Add back the scan op in its proper place.
+				ExecutionPlan_ReplaceOp(match_branch_plan, argument, parent);
+				ExecutionPlan_AddOp(parent, argument);
+				// Remove modifies from argument op since no modifications are required in this branch.
+				array_free(argument->modifies);
+				argument->modifies = NULL;
+			}
+		}
+	}
+
+	// NULL-set variables shared between the match_branch_plan and the overall plan.
+	match_branch_plan->root = NULL;
+	match_branch_plan->record_map = NULL;
+	// Free the temporary plan.
+	ExecutionPlan_Free(match_branch_plan);
+
+	return branch_match_root;
+}
+
+// Swap operation on operation children.
+static void _OpBaseSwapChildren(OpBase *op, int a, int b) {
+	assert(a >= 0 && b >= 0 && a < op->childCount && b < op->childCount);
+	OpBase *tmp = op->children[a];
+	op->children[a] = op->children[b];
+	op->children[b] = tmp;
+}
+
+static void _CreateBoundBranch(OpBase *op, ExecutionPlan *plan) {
+	rax *bound_vars = NULL;
+	const char **arguments = NULL;
+	if(plan->root) {
+		bound_vars = raxNew();
+		// Rather than cloning the record map, collect the bound variables along with their
+		// parser-generated constant strings.
+		ExecutionPlan_BoundVariables(plan->root, bound_vars);
+		// Prepare the variables for populating the Argument ops we will build.
+		arguments = ExecutionPlan_BuildArgumentModifiesArray(bound_vars);
+	}
+	OpBase *arg = NewArgumentOp(plan, arguments);
+
+	/* In case of Apply operations, we need to append, the new argument branch and
+	 * set it as the first child. */
+	if(op->type & (OPType_APPLY_MULTIPLEXER | OPType_SEMI_APPLY)) {
+		ExecutionPlan_AddOp(op, arg);
+		_OpBaseSwapChildren(op, 0, op->childCount - 1);
+	} else {
+		// -1 & everything is true;
+		op = ExecutionPlan_LocateFirstOp(op, -1);
```

Comment:
Isn't this always going to return itself?

Reply:
true. fixed